snuba/query/processors/arrayjoin_optimizer.py

```diff
@@ -0,0 +1,300 @@
+        arrayjoin_pattern = FunctionCall(
+            None,
+            String("arrayJoin"),
+            (
+                Column(
+                    None,
+                    None,
+                    Param("col", Or([String("tags.key"), String("tags.value")])),
+                ),
+            ),
+        )
+
+        arrayjoins_in_query = set()
+        for e in query.get_all_expressions():
+            match = arrayjoin_pattern.match(e)
+            if match is not None:
+                arrayjoins_in_query.add(match.string("col"))
+
+        filtered_tags = [
+            LiteralExpr(None, tag_name) for tag_name in get_filtered_tag_keys(query)
+        ]
+
+        def replace_expression(expr: Expression) -> Expression:
+            """
+            Applies the appropriate optimization on a single arrayJoin expression.
+            """
+            match = arrayjoin_pattern.match(expr)
+            if match is None:
+                return expr
+
+            if arrayjoins_in_query == {"tags.key", "tags.value"}:
+                # Both tags_key and tags_value expressions present int the query.
+                # do the arrayJoin on key-value pairs instead of independent
+                # arrayjoin for keys and values.
+                array_index = (
+                    LiteralExpr(None, 1)
+                    if match.string("col") == "tags.key"
+                    else LiteralExpr(None, 2)
+                )
+
+                if not filtered_tags:
+                    return _unfiltered_tag_pairs(expr.alias, array_index)
+                else:
+                    return _filtered_tag_pairs(expr.alias, filtered_tags, array_index)
+
+            elif filtered_tags:
+                # Only one between tags_key and tags_value is present, and
+                # it is tags_key since we found filtered tag keys.
+                return _filtered_tag_keys(expr.alias, filtered_tags)
+            else:
+                # No viable optimization
+                return expr
+
+        query.transform_expressions(replace_expression)
+
+
+def _unfiltered_tag_pairs(alias: Optional[str], array_index: LiteralExpr) -> Expression:
+    # (arrayJoin(
+    #   arrayMap((x,y) -> [x,y], tags.key, tags.value)
+    #  as all_tags)[1]
+    return arrayElement(
+        alias,
+        array_join(
+            "all_tags",
+            zip_columns(
+                ColumnExpr(None, None, "tags.key"),
+                ColumnExpr(None, None, "tags.value"),
+            ),
+        ),
+        array_index,
+    )
+
+
+def _filtered_tag_pairs(
+    alias: Optional[str],
+    filtered_tags: Sequence[LiteralExpr],
+    array_index: LiteralExpr,
+) -> Expression:
+    # (arrayJoin(arrayFilter(
+    #       pair -> arrayElement(pair, 1) IN (tags),
+    #       arrayMap((x,y) -> [x,y], tags.key, tags.value)
+    #  )) as all_tags)[1]
+    return arrayElement(
+        alias,
+        array_join(
+            "all_tags",
```

Comment:
What happens if someone actually uses the alias "all_tags" in the query?

Reply:
Sadly the same as what happens today: https://github.com/getsentry/snuba/blob/master/snuba/datasets/tags_column_processor.py#L306-L309
This same alias is already being applied in the processing code today. You would have alias shadowing.

I think, now with the ast, we could ensure the alias is not already used before when applying it.

Reply:
Sure. I think adding that check before picking a name, adding some logging, or even just renaming to `__all_tags` (or something that makes the possibility of collision clearer) would all be fine.

Reply:
Reserving a `snuba:` namespace these sorts of optimizations might not be a bad idea.

Reply:
added the prefix, will put something on the backlog to enforce `snuba:` cannot be used by the client